### PR TITLE
First version of tracking for MID

### PR DESCRIPTION
**Pull request by dstocco**
This pull-request consists of 4 closely related commits:
- the first commit defines a track for MID
- the second and third define some tools which allow for testing the tracking. In particular they add:
  * a class to find the impact point of the (generated) track on a chamber
  * a fast track generator
- the last commit is the tracking device itself with the corresponding tests.

Although they are closely related, it would be good if they can be merged without squashing since they are "atomic" commits.

### CODE REVIEW

**Comment by shahor02** [comment]
Hi @dstocco , would you consider to add ClassDefNV to classes definitions, at least to persistent ones?
It improves IO and also allows streaming the objects to debug trees (Common/Utils/test/testTreeStream.cxx)

**Comment by dstocco** [comment]
Hi Ruben,
actually, the objects that I produce are not thought to be persistent. That's why I do not need a dictionary for the classes at this level.
The idea is to run the MID tracking at FLP level, and combine the MID track (that of course contains no info on the momentum) with the future MCH one at EPN level, and that will be our persistent object.

Cheers,
Diego

**Comment by shahor02** [comment]
Hi Diego,

But I guess till you implement the FLP/EPN chain we will use the root files for the exchange (at least this is what we do for other detectors). Also, as far as I understand the root internals, if the object will be part (or base) of other (final persistent) object, it still will profit from the extra dictionary entries added by ClassDefNV (i.e. adding it only to the final object will not be enough).
Anyway, this is not obligatory (though w/o this you will not be able to profit from the debug streamers), if you want to go on w/o this, fine for me.

Cheers,
 Ruben

**Comment by shahor02** [comment]
Hi Diego,
Fine for me.
Cheers,
 Ruben

**Comment by matthiasrichter** [comment]
Hi Diego,
`TestingTools` seems to be fine. As for the other comment, just take it as load thinking what would be possible to be factored out into some common code.

Ciao Matthias

File: DataFormats/Detectors/MUON/MID/include/DataFormatsMID/Track.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+  enum class CovarianceParamIndex : int {
+    VarX,       ///< Variance on X position
+    VarY,       ///< Variance on Y position
+    VarSlopeX,  ///< Variance on X slope
+    VarSlopeY,  ///< Variance on Y slope
+    CovXSlopeX, ///< Covariance on X position and slope
+    CovYSlopeY  ///< Covariance on Y position and slope
+  };
+
```
[the remark this answers is not in the document]
Comment (aphecetche) [reply DataFormats/Detectors/MUON/MID/include/DataFormatsMID/Track.h]: Hi Diego, 

As an alternative to TestingTools (which is not completely wrong as a name but extremely generic ;-) ), one may consider FakeSimulation maybe ? It clearly states this is not a physics-grade simulation, but is more narrow in scope than "TestingTools" ? 

Just my 2cents,

[the remark this answers is not in the document]
Comment (dstocco) [reply DataFormats/Detectors/MUON/MID/include/DataFormatsMID/Track.h]: Hi 
I see your point on narrowing the scope. However, FakeSimulation does not seems correct to me. The simulation of the detector response can be "real" even if the generator is not physical.
Maybe I can restrict the scope by renaming to:
`TestingSimTools`

Cheers,
Diego

**Comment by dstocco** [comment]
Hi @matthiasrichter @aphecetche ,
I updated the code: the naming should be fine now.
However, it seems that there are currently unrelated problems in compiling O2.
@dberzano : do you know what's going wrong? I see that alibuild itself opened an issue (#965)....

Thanks,
cheers,
Diego

**Comment by dberzano** [comment]
We are currently unable to compile O2 with the latest dev of FairRoot. Please ignore the problem.

**Comment by dberzano** [comment]
Hi @matthiasrichter, there's no log file, it's not a real problem (actually: it was green until some time ago, then it turned red when rechecking with no good reason, and if you look at it now it's green again).

**Pull request merged**